### ISSUE TITLE
Issue: The crawler gets stuck on certain pages

### ISSUE THREAD

**Issue by devinat1** [janreges/siteone-crawler#23]
When I try to crawl https://www.cnn.com/2024/10/05/weather/tropical-storm-milton-florida-gulf-of-mexico/index.html, the crawler gets stuck with this output until I exit via Ctrl + C.
```

 ####                ####             #####        
 ####                ####           #######        
 ####      ###       ####         #########        
 ####     ######     ####       ###### ####        
  ######################       #####   ####        
    #######    #######       #####     ####        
    #######    #######         #       ####        
  ######################               ####        
 ####     ######     ####              ####        
 ####       ##       ####              ####        
 ####                ####       ################## 
 ####                ####       ################## 

==================================================
# SiteOne Crawler, v1.0.8.20240824               #
# Author: jan.reges@siteone.cz                   #
==================================================


Progress report           | URL                                                                       | Status | Type     | Time   | Size    | Access.  | Best pr.
-------------------------------------------------------------------------------------------------------------------------------------------------------------------
```

This is the command I used:
crawler --url=https://www.cnn.com/2024/10/05/weather/tropical-storm-milton-florida-gulf-of-mexico/index.html --offline-export-dir=/Users/devinat1/Desktop/agent-collector/utils/website-scraper/../../data/synthetic/clones/cnn --workers=10 --max-visited-urls=500 --allowed-domain-for-external-files=* --ignore-robots-txt

**Comment by devinat1** [comment]
Upon pressing ctrl + C I get the following output:
```

 ####                ####             #####        
 ####                ####           #######        
 ####      ###       ####         #########        
 ####     ######     ####       ###### ####        
  ######################       #####   ####        
    #######    #######       #####     ####        
    #######    #######         #       ####        
  ######################               ####        
 ####     ######     ####              ####        
 ####       ##       ####              ####        
 ####                ####       ################## 
 ####                ####       ################## 

==================================================
# SiteOne Crawler, v1.0.8.20240824               #
# Author: jan.reges@siteone.cz                   #
==================================================


Progress report           | URL                                                                       | Status | Type     | Time   | Size    | Access.  | Best pr.
-------------------------------------------------------------------------------------------------------------------------------------------------------------------
^C
Redirected URLs
---------------

No redirects found.

Traceback (most recent call last):
  File "/opt/miniconda3/envs/collector/lib/python3.11/asyncio/runners.py", line 118, in run

404 URLs
--------

No 404 URLs found.

    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/miniconda3/envs/collector/lib/python3.11/asyncio/base_events.py", line 654, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/Users/devinat1/Desktop/agent-collector/utils/website-scraper/siteone-scraper.py", line 70, in run_parallel_crawls
    tasks = [wrapper(command) for command in commands]
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/devinat1/Desktop/agent-collector/utils/website-scraper/siteone-scraper.py", line 67, in wrapper
    dynamic_workers = get_dynamic_workers(min_workers=min_workers, max_workers=max_workers)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/devinat1/Desktop/agent-collector/utils/website-scraper/siteone-scraper.py", line 20, in run_crawler
    process = await asyncio.create_subprocess_exec(*command)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/miniconda3/envs/collector/lib/python3.11/asyncio/subprocess.py", line 201, in communicate
    await self.wait()
  File "/opt/miniconda3/envs/collector/lib/python3.11/asyncio/subprocess.py", line 137, in wait
    return await self._transport._wait()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/miniconda3/envs/collector/lib/python3.11/asyncio/base_subprocess.py", line 230, in _wait
    return await waiter
           ^^^^^^^^^^^^
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/devinat1/Desktop/agent-collector/utils/website-scraper/siteone-scraper.py", line 78, in <module>
    
    ^
  File "/opt/miniconda3/envs/collector/lib/python3.11/asyncio/runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/opt/miniconda3/envs/collector/lib/python3.11/asyncio/runners.py", line 123, in run
    raise KeyboardInterrupt()
KeyboardInterrupt

(collector) devinat1@Devins-MacBook-Pro website-scraper % 
(collector) devinat1@Devins-MacBook-Pro website-scraper % 
(collector) devinat1@Devins-MacBook-Pro website-scraper % 
SSL/TLS info
------------

Info                   | Text                                                                                                                                    
-------------------------------------------------------------------------------------------------------------------------------------------------------------------
Issuer                 | C = BE, O = GlobalSign nv-sa, CN = GlobalSign Atlas R3 DV TLS CA 2024 Q3                                                                
Subject                | CN = cnn.com                                                                                                                            
Valid from             | Sep 27 19:18:30 2024 GMT (VALID already 8 day(s))                                                                            
Valid to               | Oct 29 19:18:29 2025 GMT (VALID still for 389 day(s))                                                                        
Supported protocols    | TLSv1.2, TLSv1.3                                                                                                  
RAW certificate output | Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            01:cc:f1…f:c9:ba:78:c1:d5:68:10:
        09:8a:92:fa
RAW protocols output   | depth=2 OU = GlobalSign Root CA - R3, O = GlobalSign, CN = GlobalSign
verify return:1
depth…: no
    Max Early Data: 0
---
read R BLOCK


TOP fastest URLs
----------------

No fast URLs fastest than 1 second(s) found.


TOP slowest URLs
----------------

No slow URLs slowest than 0.01 second(s) found.


SEO metadata
------------

No URLs.


OpenGraph metadata
------------------

No URLs with OpenGraph data (og:* or twitter:* meta tags).


Heading structure
-----------------

No URLs to analyze heading structure.


HTTP headers
------------

No HTTP headers found.


HTTP header values
------------------

No HTTP headers found.


Best practices
--------------

Analysis name                  | OK    | Notice | Warning | Critical
----------------------------------------------------------------------
Title uniqueness (> 10%)       | 0     | 0      | 1       | 0       
Description uniqueness (> 10%) | 0     | 0      | 1       | 0       
Brotli support                 | 0     | 0      | 0       | 0       
WebP support                   | 0     | 0      | 1       | 0       
AVIF support                   | 0     | 0      | 1       | 0       


Accessibility
-------------

Nothing to report.


Source domains
--------------

No source domains found.


Content types
-------------

No URLs found.


Content types (MIME types)
--------------------------

No MIME types found.


DNS info
--------

DNS resolving tree                                                    
------------------------------------------------------------------------
www.cnn.com                                                           
  cnn-tls.map.fastly.net                                              
    IPv4: 151.101.131.5                                               
    IPv4: 151.101.195.5                                               
    IPv4: 151.101.3.5                                                 
    IPv4: 151.101.67.5                                                
                                                                      
DNS server: 156.154.54.16                                             


Security
--------

Nothing to report.


Analysis stats
--------------

Class::method                                        | Exec time | Exec count
-------------------------------------------------------------------------------
SslTlsAnalyzer::getTLSandSSLCertificateInfo          | 500 ms    | 1         
SeoAndOpenGraphAnalyzer::analyzeSeo                  | 0 ms      | 1         
BestPracticeAnalyzer::checkBrotliSupport             | 0 ms      | 1         
SeoAndOpenGraphAnalyzer::analyzeOpenGraph            | 0 ms      | 1         
SeoAndOpenGraphAnalyzer::analyzeHeadings             | 0 ms      | 1         
BestPracticeAnalyzer::checkTitleUniqueness           | 0 ms      | 1         
BestPracticeAnalyzer::checkWebpSupport               | 0 ms      | 1         
BestPracticeAnalyzer::checkMetaDescriptionUniqueness | 0 ms      | 1         
BestPracticeAnalyzer::checkAvifSupport               | 0 ms      | 1         


Content processor stats
-----------------------

No content processors found.


======================================================================================================================================================================
Total execution time 107 s using 10 workers and 2048M memory limit (max used 6 MB)
Total of 0 visited URLs with a total size of 0 B and power of 0 reqs/s with download speed 0 B/s
Response times: AVG 0 ms MIN 0 ms MAX 0 ms TOTAL 0 ms
======================================================================================================================================================================

Summary
-------

⚠️ No titles provided for uniqueness check.
⚠️ No meta descriptions provided for uniqueness check.
⚠️ No WebP image found on the website.
⚠️ No AVIF image found on the website.
⏩ DNS IPv6: domain www.cnn.com does not support IPv6 (DNS server: 156.154.54.16).
✅ Redirects - no redirects found.
✅ 404 OK - all pages exists, no non-existent pages found.
✅ SSL/TLS certificate is valid until Oct 29 19:18:29 2025 GMT. Issued by C = BE, O = GlobalSign nv-sa, CN = GlobalSign Atlas R3 DV TLS CA 2024 Q3. Subject is CN = cnn.com.
✅ SSL/TLS certificate issued by 'C = BE, O = GlobalSign nv-sa, CN = GlobalSign Atlas R3 DV TLS CA 2024 Q3'.
✅ Performance OK - all non-media URLs are faster than 3 seconds.
✅ HTTP headers - found 0 unique headers.
✅ All pages support Brotli compression.
✅ All pages have quoted attributes.
(collector) devinat1@Devins-MacBook-Pro website-scraper % python siteone-scraper.py
['crawler', '--url=https://www.cnn.com/2024/10/05/weather/tropical-storm-milton-florida-gulf-of-mexico/index.html', '--offline-export-dir=/Users/devinat1/Desktop/agent-collector/utils/website-scraper/../../data/synthetic/clones/cnn', '--workers=10', '--max-visited-urls=500', '--allowed-domain-for-external-files=*', '--ignore-robots-txt']

 ####                ####             #####        
 ####                ####           #######        
 ####      ###       ####         #########        
 ####     ######     ####       ###### ####        
  ######################       #####   ####        
    #######    #######       #####     ####        
    #######    #######         #       ####        
  ######################               ####        
 ####     ######     ####              ####        
 ####       ##       ####              ####        
 ####                ####       ################## 
 ####                ####       ################## 

==================================================
# SiteOne Crawler, v1.0.8.20240824               #
# Author: jan.reges@siteone.cz                   #
==================================================


Progress report           | URL                                                                       | Status | Type     | Time   | Size    | Access.  | Best pr.
-------------------------------------------------------------------------------------------------------------------------------------------------------------------
^CTraceback (most recent call last):
  File "/opt/miniconda3/envs/collector/lib/python3.11/asyncio/runners.py", line 118, in run

Redirected URLs
---------------

No redirects found.

    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/miniconda3/envs/collector/lib/python3.11/asyncio/base_events.py", line 654, in run_until_complete
    return future.result()

404 URLs
--------

No 404 URLs found.

           ^^^^^^^^^^^^^^^
  File "/Users/devinat1/Desktop/agent-collector/utils/website-scraper/siteone-scraper.py", line 71, in run_parallel_crawls
    await asyncio.gather(*tasks)
  File "/Users/devinat1/Desktop/agent-collector/utils/website-scraper/siteone-scraper.py", line 68, in wrapper
    await run_crawler(*command, workers=dynamic_workers)
  File "/Users/devinat1/Desktop/agent-collector/utils/website-scraper/siteone-scraper.py", line 21, in run_crawler
    await process.communicate()
  File "/opt/miniconda3/envs/collector/lib/python3.11/asyncio/subprocess.py", line 201, in communicate
    await self.wait()
  File "/opt/miniconda3/envs/collector/lib/python3.11/asyncio/subprocess.py", line 137, in wait
    return await self._transport._wait()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/miniconda3/envs/collector/lib/python3.11/asyncio/base_subprocess.py", line 230, in _wait
    return await waiter
           ^^^^^^^^^^^^
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/devinat1/Desktop/agent-collector/utils/website-scraper/siteone-scraper.py", line 79, in <module>
    asyncio.run(run_parallel_crawls(source_path, output_dir, max_concurrent_tasks=5))
  File "/opt/miniconda3/envs/collector/lib/python3.11/asyncio/runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/opt/miniconda3/envs/collector/lib/python3.11/asyncio/runners.py", line 123, in run
    raise KeyboardInterrupt()
KeyboardInterrupt

(collector) devinat1@Devins-MacBook-Pro website-scraper % 
SSL/TLS info
------------

Info                   | Text                                                                                                                                    
-------------------------------------------------------------------------------------------------------------------------------------------------------------------
Issuer                 | C = BE, O = GlobalSign nv-sa, CN = GlobalSign Atlas R3 DV TLS CA 2024 Q3                                                                
Subject                | CN = cnn.com                                                                                                                            
Valid from             | Sep 27 19:18:30 2024 GMT (VALID already 8 day(s))                                                                            
Valid to               | Oct 29 19:18:29 2025 GMT (VALID still for 389 day(s))                                                                        
Supported protocols    | TLSv1.2, TLSv1.3                                                                                                  
RAW certificate output | Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            01:cc:f1…f:c9:ba:78:c1:d5:68:10:
        09:8a:92:fa
RAW protocols output   | depth=2 OU = GlobalSign Root CA - R3, O = GlobalSign, CN = GlobalSign
verify return:1
depth…: no
    Max Early Data: 0
---
read R BLOCK


TOP fastest URLs
----------------

No fast URLs fastest than 1 second(s) found.


TOP slowest URLs
----------------

No slow URLs slowest than 0.01 second(s) found.


SEO metadata
------------

No URLs.


OpenGraph metadata
------------------

No URLs with OpenGraph data (og:* or twitter:* meta tags).


Heading structure
-----------------

No URLs to analyze heading structure.


HTTP headers
------------

No HTTP headers found.


HTTP header values
------------------

No HTTP headers found.


Best practices
--------------

Analysis name                  | OK    | Notice | Warning | Critical
----------------------------------------------------------------------
Title uniqueness (> 10%)       | 0     | 0      | 1       | 0       
Description uniqueness (> 10%) | 0     | 0      | 1       | 0       
Brotli support                 | 0     | 0      | 0       | 0       
WebP support                   | 0     | 0      | 1       | 0       
AVIF support                   | 0     | 0      | 1       | 0       


Accessibility
-------------

Nothing to report.


Source domains
--------------

No source domains found.


Content types
-------------

No URLs found.


Content types (MIME types)
--------------------------

No MIME types found.


DNS info
--------

DNS resolving tree                                                    
------------------------------------------------------------------------
www.cnn.com                                                           
  cnn-tls.map.fastly.net                                              
    IPv4: 151.101.67.5                                                
    IPv4: 151.101.131.5                                               
    IPv4: 151.101.195.5                                               
    IPv4: 151.101.3.5                                                 
                                                                      
DNS server: 156.154.54.16                                             


Security
--------

Nothing to report.


Analysis stats
--------------

Class::method                                        | Exec time | Exec count
-------------------------------------------------------------------------------
SslTlsAnalyzer::getTLSandSSLCertificateInfo          | 517 ms    | 1         
SeoAndOpenGraphAnalyzer::analyzeSeo                  | 0 ms      | 1         
SeoAndOpenGraphAnalyzer::analyzeHeadings             | 0 ms      | 1         
SeoAndOpenGraphAnalyzer::analyzeOpenGraph            | 0 ms      | 1         
BestPracticeAnalyzer::checkBrotliSupport             | 0 ms      | 1         
BestPracticeAnalyzer::checkTitleUniqueness           | 0 ms      | 1         
BestPracticeAnalyzer::checkWebpSupport               | 0 ms      | 1         
BestPracticeAnalyzer::checkMetaDescriptionUniqueness | 0 ms      | 1         
BestPracticeAnalyzer::checkAvifSupport               | 0 ms      | 1         


Warning: opendir(/Users/devinat1/Desktop/agent-collector/utils/website-scraper/../../data/synthetic/clones/cnn): Failed to open directory: No such file or directory in /usr/local/siteone-crawler/src/Crawler/Utils.php on line 779

Content processor stats
-----------------------

No content processors found.


======================================================================================================================================================================
Total execution time 258 s using 10 workers and 2048M memory limit (max used 6 MB)
Total of 0 visited URLs with a total size of 0 B and power of 0 reqs/s with download speed 0 B/s
Response times: AVG 0 ms MIN 0 ms MAX 0 ms TOTAL 0 ms
======================================================================================================================================================================

Summary
-------

⚠️ No titles provided for uniqueness check.
⚠️ No meta descriptions provided for uniqueness check.
⚠️ No WebP image found on the website.
⚠️ No AVIF image found on the website.
⏩ DNS IPv6: domain www.cnn.com does not support IPv6 (DNS server: 156.154.54.16).
✅ Redirects - no redirects found.
✅ 404 OK - all pages exists, no non-existent pages found.
✅ SSL/TLS certificate is valid until Oct 29 19:18:29 2025 GMT. Issued by C = BE, O = GlobalSign nv-sa, CN = GlobalSign Atlas R3 DV TLS CA 2024 Q3. Subject is CN = cnn.com.
✅ SSL/TLS certificate issued by 'C = BE, O = GlobalSign nv-sa, CN = GlobalSign Atlas R3 DV TLS CA 2024 Q3'.
✅ Performance OK - all non-media URLs are faster than 3 seconds.
✅ HTTP headers - found 0 unique headers.
✅ All pages support Brotli compression.
✅ All pages have quoted attributes.
✅ All pages have inline SVGs smaller than 5120 bytes.
✅ All pages have inline SVGs with less than 5 duplicates.
✅ All pages have valid or none inline SVGs.
✅ All pages without multiple <h1> headings.
✅ All pages have <h1> heading.
✅ All pages have heading structure without skipped levels.
✅ All pages have DOM depth less than 30.
✅ All pages have clickable (interactive) phone numbers.
✅ All pages have valid HTML.
✅ All pages have image alt attributes.
✅ All pages have form labels.
✅ All pages have aria labels.
✅ All pages have role attributes.
✅ All pages have lang attribute.
✅ DNS IPv4 OK: domain www.cnn.com resolved to 151.101.67.5, 151.101.131.5, 151.101.195.5, 151.101.3.5 (DNS server: 156.154.54.16).
✅ Security - no findings.
📌 DNS Aliases: IP(s) for domain www.cnn.com were resolved by CNAME chain www.cnn.com > cnn-tls.map.fastly.net.
📌 Text report saved to '/usr/local/siteone-crawler/tmp/www.cnn.com.output.20241005-200254.txt' and took 1 ms.
📌 JSON report saved to '/usr/local/siteone-crawler/tmp/www.cnn.com.output.20241005-200254.json' and took 0 ms.
📌 HTML report saved to '/usr/local/siteone-crawler/tmp/www.cnn.com.report.20241005-200254.html' and took 5 ms.
📌 Offline website generated to '/Users/devinat1/Desktop/agent-collector/utils/website-scraper/../../data/synthetic/clones/cnn' and took 0 ms.


Fatal error: Uncaught Swoole\ExitException: I caught the manual stop of the script. Therefore, the statistics only contain processed URLs until the script stops. in /usr/local/siteone-crawler/src/Crawler/Crawler.php:173
Stack trace:
#0 [internal function]: Crawler\Crawler->Crawler\{closure}(2)
#1 {main}
  thrown in /usr/local/siteone-crawler/src/Crawler/Crawler.php on line 173
  ```

**Comment by janreges** [comment]
Hi @devinat1,

I used your command exactly, I just modified the path where to save the exported web. It works without any problem (see screenshot).

Have you tried it again? It is possible that you have encountered some limitations with CNN and their CDN or WAF.

Since you are crawling several sites in parallel with a high number of workers through your own python script, it is possible that you have encountered some limits of your ISP or even your computer - I don't mean performance limits, but limits on the number of open TCP connections, file descriptors, etc.

<img width="868" alt="image" src="https://github.com/user-attachments/assets/6cefe922-6593-4e69-95c7-560468a9b3d7">
